### PR TITLE
Add `chainerx.cumsum`

### PR DESCRIPTION
**Pull request by aksub99**
Add `chainerx.cumsum`.

### CODE REVIEW

File: chainerx_cc/chainerx/kernels/indexing.h [hunk 1]
```diff
@@ -50,4 +50,11 @@ class WhereASSKernel : public Kernel {
     virtual void Call(const Array& condition, Scalar x, Scalar y, const Array& out) = 0;
 };
 
+class CumsumKernel : public Kernel {
```
Comment (asi1024) [inline chainerx_cc/chainerx/kernels/indexing.h]: This kernel should be declared in `kernels/reduction.h`.

Comment (aksub99) [reply chainerx_cc/chainerx/kernels/indexing.h]: Ok. Will move it to `reduction.h`. I assumed that since the function doesn't reduce the dimensions of the array,  it could be added to indexing.

Comment (asi1024) [reply chainerx_cc/chainerx/kernels/indexing.h]: NumPy classifies `cumsum` same routine as `sum` and `prod`.

Comment (aksub99) [reply chainerx_cc/chainerx/kernels/indexing.h]: I have used `MakeRollingPermutation` (defined in `indexing.cc`) in the definition of `cumsum`. It wouldn't be appropriate to redefine it in `reduction.cc` right?

File: chainerx_cc/chainerx/cuda/cuda_device/reduction.cu [hunk 2]
```diff
@@ -131,6 +131,15 @@ public:
 
 CHAINERX_CUDA_REGISTER_KERNEL(SumKernel, CudaSumKernel);
 
+class CudaCumsumKernel : public CumsumKernel {
+public:
+    void Call(const Array& a, int8_t axis, const Array& out) override {
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/reduction.cu]: Could you comment out these arguments to avoid unused-parameter error?
```suggestion
    void Call(const Array& /* a */, int8_t /* axis */, const Array& /* out */) override {
```

Comment (aksub99) [reply chainerx_cc/chainerx/cuda/cuda_device/reduction.cu]: Done.

**Comment by aksub99** [comment]
I got rid of the `chainerx._testing` travis failure. The native implementation of `cumsum` still doesn't work as expected. Will make the fixes ASAP.

**Comment by aksub99** [comment]
Fixed the implementation (Seems to work correctly when I tested locally). Only tests are remaining.

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py [hunk 3]
```diff
@@ -247,3 +259,27 @@ def test_log_softmax_invalid(device, a_shape, axis, dtype):
     a = array_utils.create_dummy_ndarray(chainerx, a_shape, dtype)
     with pytest.raises(chainerx.DimensionError):
         return chainerx.log_softmax(a, axis=axis)
+
+
+@op_utils.op_test(['native:0'])
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: Could you add TODO about CUDA device implementation?

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py [hunk 4]
```diff
@@ -247,3 +259,27 @@ def test_log_softmax_invalid(device, a_shape, axis, dtype):
...
+@chainer.testing.parameterize_pytest(
+    'in_dtypes,out_dtype', math_utils.in_out_dtypes_math_functions)
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: `chainerx.cumsum` should use the same dtype-rule as that of `chainerx.sum`.
https://github.com/chainer/chainer/blob/bc3dbef7eec8b77c72a7844844ee2401e1b026e1/tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py#L54-L67.

Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: How about defining this list as `_in_out_dtypes_sum` and use it?

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py [hunk 5]
```diff
@@ -247,3 +259,27 @@ def test_log_softmax_invalid(device, a_shape, axis, dtype):
...
+@chainer.testing.parameterize_pytest('shape,axis', _cumsum_params)
+class TestCumsum(math_utils.UnaryMathTestBase, op_utils.NumpyOpTest):
+
+    input = 'random'
+
+    def setup(self):
+        super().setup()
+        if self.in_dtypes == 'float16':
+            # TODO(imanishi): Support device implementation and remove this.
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: This comment is for `chainerx.logsumexp`. Please remove this line.

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py [hunk 6]
```diff
@@ -247,3 +259,27 @@ def test_log_softmax_invalid(device, a_shape, axis, dtype):
...
+            self.check_forward_options.update({'rtol': 3e-3, 'atol': 3e-3})
+
+    def forward_xp(self, inputs, xp):
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: How about defining `func` method instead of `forward_xp`? (Please refer `TestSum`)

**Comment by aksub99** [comment]
The tests are failing for `float16` and `bool` inputs.

File: chainerx_cc/chainerx/native/native_device/reduction.cc [hunk 7]
```diff
@@ -106,6 +106,73 @@ class NativeSumKernel : public SumKernel {
...
+
+                for (it_left.Restart(); it_left; ++it_left) {
+                    it_out.CopyIndex(it_left);
+                    it_prev.CopyIndex(it_left);
+
+                    for (it_right.Restart(); it_right; ++it_right) {
+                        it_out.CopyIndex(it_right, it_left.ndim() + it.ndim());
+                        it_prev.CopyIndex(it_right, it_left.ndim() + it_axis.ndim());
+                        it_prev.index()[axis] -= 1;
+                        out_iarray[it_out] += out_iarray[it_prev];
```
Comment (asi1024) [inline chainerx_cc/chainerx/native/native_device/reduction.cc]: float16 arrays store their items as `uint16_t` values because C++ and CUDA don't have the float16 type as a primitive type.
https://github.com/chainer/chainer/blob/c47a3785bca0e36af35406f43776004ad1030da1/chainerx_cc/chainerx/dtype.h#L108
So, before calculating summation, we need to convert types with `native_internal::StorageToDataType` defined in `chainerx/native/data_type.h`.
```suggestion
                        native_internal::StorageToDataType<T>(out_iarray[it_out]) +=
                                native_internal::StorageToDataType<T>(out_iarray[it_prev]);

```

**Comment by asi1024** [comment]
`numpy.cumsum` returns `int64` ndarray when bool or integer ndarray has taken as an input. `chainerx.cumsum` should also follow this dtype rule.

File: chainerx_cc/chainerx/native/native_device/reduction.cc (hunk 7 above)
Comment (aksub99) [reply chainerx_cc/chainerx/native/native_device/reduction.cc]: Thanks for the explanation.

**Comment by aksub99** [comment]
Have made the suggested changes.

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py [hunk 8]
```diff
@@ -38,8 +38,18 @@
 ]
 
 
-@op_utils.op_test(['native:0', 'cuda:0'])
-@chainer.testing.parameterize_pytest('in_dtypes,out_dtype', [
+_cumsum_params = [
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: Add test for `axis=None`.

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py [hunk 9]
```diff
@@ -247,3 +262,27 @@ def test_log_softmax_invalid(device, a_shape, axis, dtype):
...
+        super().setup()
+        in_dtype, = self.in_dtypes
+        if in_dtype == 'float16':
+            self.check_forward_options.update({'rtol': 1e-2, 'atol': 1e-2})
+            self.check_backward_options.update({'rtol': 1e-2, 'atol': 1e-2})
+            self.check_double_backward_options.update(
+                {'rtol': 1e-2, 'atol': 1e-2})
+
+    def func(self, xp, a):
+        return xp.cumsum(a, axis=self.axis).astype(self.out_dtype)
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]:
```suggestion
        return xp.cumsum(a, axis=self.axis)
```

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py (hunk 8 above)
Comment (aksub99) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: The current implementation doesn't support `axis=None` yet. In the native implementation, I use the `int` value of `axis` and have hence not defined `axis` to be of `OptionalAxes` type. Is there any way I could make it support `None` without modifying the type of `axis` argument?

Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: Please modify the type of `axis` argument.

Comment (aksub99) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: If I convert the type of the `axis` argument to `Axes`, how could I access the `int` value of it? I looked at the methods in https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/axes.h but wasn't able to find any such way. Would you have any suggestions?

Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: Please use `absl::optional<int8_t>` as the type of `axis`.

Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: `CumsumKernel` is OK to take `int8_t axis`.

File: chainerx_cc/chainerx/native/native_device/reduction.cc (hunk 7 above)
Comment (asi1024) [reply chainerx_cc/chainerx/native/native_device/reduction.cc]: Please include `data_type.h`

File: tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py (hunk 8 above)
Comment (aksub99) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: With `axis` being `int8_t` type, I am not able to make use of the `GetSortedAxesOrAll` method that works with only `Axes` type. So, I get the default behavior(when `axis` argument is not provided) as `axis=0` instead of `axis=all`.

Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: `GetSortedAxesOraAll` seems not needed in this case because we don't need sorted axes.

Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: As temporary support, how about fixing to reshape the input array to 1-dim?

Comment (aksub99) [reply tests/chainerx_tests/unit_tests/routines_tests/test_reduction.py]: I have made the changes. Tests pass for `axis=None` now.

File: chainerx_cc/chainerx/native/native_device/reduction.cc (hunk 7 above)
Comment (aksub99) [reply chainerx_cc/chainerx/native/native_device/reduction.cc]: Done.

File: chainerx_cc/chainerx/routines/reduction.cc [hunk 10]
```diff
@@ -98,14 +98,21 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
...
-    int8_t axis_norm = internal::NormalizeAxis(axis, a.ndim());
+Array Cumsum(const Array& a, absl::optional<int8_t> axis) {
+    int8_t axis_norm;
+    Array a_reshaped = Copy(a);
+    if (axis.has_value()) {
+        axis_norm = internal::NormalizeAxis(*axis, a.ndim());
+    } else {
+        axis_norm = 0;
+        a_reshaped = a.Reshape(Shape{a.GetTotalSize()});
+    }
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/reduction.cc]: Remove the redundant copy.
```py
    Array a_reshaped{}
    if (axis.has_value()) {
        axis_norm = internal::NormalizeAxis(*axis, a.ndim());
        a_reshaped = Copy(a);
    } else {
        axis_norm = 0;
        a_reshaped = a.Reshape(Shape{a.GetTotalSize()});
    }
```

File: chainerx_cc/chainerx/native/native_device/reduction.cc (hunk 7 above)
Comment (aksub99) [reply chainerx_cc/chainerx/native/native_device/reduction.cc]: Done.

File: chainerx_cc/chainerx/routines/reduction.cc (hunk 10 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/reduction.cc]: Done.

**Comment by aksub99** [comment]
@asi1024 I have added the backward implementation. The backward tests pass but the double backward tests fail. Could you have a look at the implementation please?
I suspect the problem might be with the usage of `Cumsum` itself in the backward implementation.

**Comment by aksub99** [comment]
Have rebased with master to include bug fix in `Flip` from #7727 . The tests pass now.

File: chainerx_cc/chainerx/routines/reduction.cc [hunk 11]
```diff
@@ -98,4 +98,56 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
...
+            out_dtype = Dtype::kInt64;
+            break;
+        case DtypeKind::kUInt:
+            out_dtype = Dtype::kInt64;  // TODO(niboshi): This should be kUInt64
+            break;
+        default:
+            out_dtype = a_reshaped.dtype();
+    }
+
+    const Array& out_cast = out.AsType(out_dtype);
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/reduction.cc]: Redundant copy.
```suggestion
    const Array& out = Empty(out.shape(), out_dtype(), a_reshaped.device());
```

File: chainerx_cc/chainerx/routines/reduction.cc [hunk 12]
```diff
@@ -98,4 +98,56 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
...
+    const Array& a_cast = a_reshaped.dtype() != out_cast.dtype() ? a_reshaped.AsType(out_cast.dtype()) : a_reshaped;
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/reduction.cc]: Please move this cast to the device implementation.

File: chainerx_cc/chainerx/routines/reduction.cc [hunk 13]
```diff
@@ -98,4 +98,56 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
...
+
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<CumsumKernel>(a_cast, axis_norm, out_cast);
+    }
+
+    BackwardBuilder bb{"cumsum", a, out_cast};
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/reduction.cc]: For simplification, it seems better to give `a_cast` (or `a_reshaped`) as an argument.

File: chainerx_cc/chainerx/native/native_device/reduction.cc [hunk 14]
```diff
@@ -106,6 +107,68 @@ class NativeSumKernel : public SumKernel {
 
 CHAINERX_NATIVE_REGISTER_KERNEL(SumKernel, NativeSumKernel);
 
+class NativeCumsumKernel : public CumsumKernel {
```
Comment (asi1024) [inline chainerx_cc/chainerx/native/native_device/reduction.cc]: Please add `TODO(imanishi): Performance improvement.`

File: chainerx_cc/chainerx/routines/reduction.cc [hunk 15]
```diff
@@ -98,4 +98,56 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
...
+    if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+        bt.Define([axis, axis_norm, in_shape = a.shape()](BackwardContext& bctx) {
+            const Array& gout = *bctx.output_grad();
+            if (axis.has_value()) {
+                bctx.input_grad() = Flip(Cumsum(Flip(gout, axis_norm), axis_norm), axis_norm);
+            } else {
+                Array input_grad = Flip(Cumsum(Flip(gout, 0), 0), 0);
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/reduction.cc]: `gout` will be flipped twice during the double-backward. It can be reduced.

File: chainerx_cc/chainerx/routines/reduction.cc (hunk 11 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/reduction.cc]: Done.

File: chainerx_cc/chainerx/routines/reduction.cc (hunk 12 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/reduction.cc]: Done.

File: chainerx_cc/chainerx/routines/reduction.cc (hunk 13 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/reduction.cc]: I had originally given `a_reshaped` as an argument to `bb`, but we require `in_shape` to be `a_shape` and not `a_reshaped` since the shape of the gradient vector must be same as the input.

File: chainerx_cc/chainerx/routines/reduction.cc [hunk 16]
```diff
@@ -98,4 +98,53 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
     return x_cast - LogSumExp(x_cast, axis.has_value() ? axis : OptionalAxes{1}, true);
 }
 
+Array Cumsum(const Array& a, absl::optional<int8_t> axis) {
+    int8_t axis_norm;
+    Array a_reshaped{};
+    if (axis.has_value()) {
+        axis_norm = internal::NormalizeAxis(*axis, a.ndim());
+        a_reshaped = Copy(a);
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/reduction.cc]:
```suggestion
        a_reshaped = a;
```

File: chainerx_cc/chainerx/routines/reduction.cc (hunk 13 above)
Comment (asi1024) [reply chainerx_cc/chainerx/routines/reduction.cc]: > since the shape of the gradient vector must be same as the input.
Yes. the shape of the gradient vector (`input_grad`) must be the same as the shape of input (an argument of `bb`). So you should fix also backward implementation.

File: chainerx_cc/chainerx/native/native_device/reduction.cc (hunk 14 above)
Comment (asi1024) [reply chainerx_cc/chainerx/native/native_device/reduction.cc]: Could you move the comment up one line? (before class definition)

File: chainerx_cc/chainerx/routines/reduction.cc (hunk 13 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/reduction.cc]: Sorry, I am slightly confused. If I change to `BackwardBuilder bb{"cumsum", a_reshaped, out}`, then `input_grad` will have to be the same shape as `a_reshaped` right (Even if I modify the backward implementation)? So, in the case where the `axis=None`, `a_reshaped` will have a different shape than `a`.

Comment (aksub99) [reply chainerx_cc/chainerx/routines/reduction.cc]: Done.

File: chainerx_cc/chainerx/routines/reduction.cc [hunk 17]
```diff
@@ -131,9 +131,9 @@ Array Cumsum(const Array& a, absl::optional<int8_t> axis) {
         a.device().backend().CallKernel<CumsumKernel>(a_reshaped, axis_norm, out);
     }
 
-    BackwardBuilder bb{"cumsum", a, out};
+    BackwardBuilder bb{"cumsum", a_reshaped, out};
     if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
-        bt.Define([axis, axis_norm, in_shape = a.shape()](BackwardContext& bctx) {
+        bt.Define([axis, axis_norm, in_shape = a_reshaped.shape()](BackwardContext& bctx) {
             const Array& gout = *bctx.output_grad();
             if (axis.has_value()) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/reduction.cc]: This `if` branching seems not needed.

Comment (aksub99) [reply chainerx_cc/chainerx/routines/reduction.cc]: Thanks for pointing that out. Fixed.

**Comment by aksub99** [comment]
I have added a `TODO` to improve backward implementation.

**Comment by aksub99** [comment]
Is this ready to merge @asi1024 ? Can I go ahead with `cumprod` and `nonzero` ?

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM. Thank you for the PR!
